logger/src/lib.rs

```diff
@@ -0,0 +1,152 @@
+                std::env::temp_dir()
+            } else {
+                let path = Path::new(&driver_path);
+                path.parent()
+                    .map(|p| p.parent().map(|p| p.join("logs")).unwrap())
+                    .unwrap_or_else(std::env::temp_dir)
+            };
+
+            if let Some(log_dir_str) = log_dir.to_str() {
+                if let Ok(appender) = Self::file_appender(log_dir_str) {
+                    match Self::init_logger(appender) {
+                        Ok(handle) => Some(Self { handle, log_dir }),
+                        Err(_) => None,
+                    }
+                } else {
+                    None
+                }
+            } else {
+                None
+            }
+        }) {
+            Ok(logger) => logger,
+            Err(_) => None,
+        }
+    }
+
+    pub fn set_log_level(&self, level_filter: String) {
+        let level_filter = match level_filter.to_lowercase().as_str() {
+            "error" => LevelFilter::Error,
+            "warn" => LevelFilter::Warn,
+            "info" => LevelFilter::Info,
+            "debug" => LevelFilter::Debug,
+            "trace" => LevelFilter::Trace,
+            _ => LevelFilter::Info,
+        };
+
+        if let Some(log_dir) = self.log_dir.to_str() {
+            if let Ok(appender) = Logger::file_appender(log_dir) {
+                let config = Config::builder()
+                    .appender(Appender::builder().build("logfile", Box::new(appender)))
+                    .build(Root::builder().appender("logfile").build(level_filter))
+                    .unwrap();
+                self.handle.set_config(config);
+            }
+        }
+    }
+
+    fn file_appender(log_dir: &str) -> Result<RollingFileAppender, std::io::Error> {
+        let file_path = format!("{log_dir}/mongo_odbc.log");
+        let roller_pattern = format!("{log_dir}/mongo_odbc.log.{{}}");
+
+        let roller = FixedWindowRoller::builder()
+            .build(&roller_pattern, 10)
+            .unwrap();
+        let trigger = SizeTrigger::new(LOG_FILE_SIZE);
+        let policy = CompoundPolicy::new(Box::new(trigger), Box::new(roller));
+
+        RollingFileAppender::builder()
+            .encoder(Box::new(PatternEncoder::new(
+                "{d(%Y-%m-%d %H:%M:%S)(utc)} - {h({l})}: {m}{n}",
+            )))
+            .append(true)
+            .build(file_path, Box::new(policy))
+    }
+
+    fn init_logger(logfile: RollingFileAppender) -> Result<Handle, ()> {
+        let config = Config::builder()
+            .appender(Appender::builder().build("logfile", Box::new(logfile)))
+            .build(Root::builder().appender("logfile").build(LevelFilter::Info))
+            .map_err(|_e| ())?;
+        log4rs::init_config(config).map_err(|_e| ())
+    }
+}
+
+#[cfg(test)]
+mod driver {
+
+    use std::{fs, path::Path};
+
+    use super::*;
+    use log::{debug, error, info};
+
+    #[test]
+    fn logger() {
```

Comment:
Why do the 3 log levels we use (`info`, `debug`, `error`) result in the same lines being logged? Aren't the filters supposed to filter out certain log messages? Like `debug` includes everything, `info` omits `debug!` logs, and `error` omits `debug!` and `info!` logs? I may be misunderstanding this.

Reply:
Logs are hierarchical. error level is the highest we use, which filters out info and debug. Likewise, info will filter out debug, but not error. Debug includes everything above it. This way we still see error messages when the loglevel is at info, for example.

Reply:
Oh my goodness I entirely missed the `!`s in the `assert!`s. That's why I was so confused. I read this as all three log levels including all 3 log lines every time.

Reply:
and I misread your comment!